### PR TITLE
Leaderboard sorting and pagination 

### LINKED ISSUE #63: Leaderboard sorting
Add sorting with dropdown for leaderboard page.

image

Tentatively, have the options like this. Can also discuss other optinos.
image

### PR DESCRIPTION
**Pull request by SheepTester**
# Info

Closes #63.

# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

I added a fancier dropdown and pagination.

I copied the pagination code from the ACM dev team take-home btw 

## Changes

- added custom dropdown
  - it has past week etc options
- it filters by year
  - unlike the current website, filtering by academic year is a bit more dumb in this implementation. The current(old) website imports a npm package maintained by ACM that hard codes the quarter start and end dates to determine the time range for a school year. Here, I just assumed the school years change on August 1
- I copied my pagination code from the take-home assignment so loading the all time leaderboard no longer crashes my browser
- the my position button disables itself when you are not on the leaderboard

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) 

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [ ] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [x] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Please include a screenshot of your Cypress testing suite passing successfully.

If you made any visual changes to the website, please include relevant screenshots below.

Screenshot_20231014-155355

### CODE REVIEW

File: src/components/leaderboard/PaginationControls/index.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        disabled={page <= 0}
+      >
+        <LeftArrowIcon />
+      </button>
+      <div className={style.paginationText}>
+        <input
+          className={style.pageNumber}
+          type="text"
+          inputMode="numeric"
+          pattern="[0-9]*"
```
Comment (farisashai) [inline src/components/leaderboard/PaginationControls/index.tsx]: might be useful to restrict negative input values and reset to 1. similarly for values greater than the max page like if they type 100 and hit enter, update the input to the max value

**Review by farisashai — COMMENTED** [review]
Can you double check that the clickable region is the entire dropdown selection? When I click directly on the up/down arrow svgs, nothing is happening but it shows a pointer cursor so either the onClick is only defined for the select tag or the svg is eating the click event.

image

**Review by farisashai — APPROVED** [review]
feature looks good to me
2 small niits
- time period dropdown has no border which looks a little weird with contrast on the page and indicating it's clickable
- default leaderbord to all time might be better (avoid no results view)
image

**Comment by SheepTester** [comment]
> * default leaderbord to all time might be better (avoid no results view)

This shouldn't happen on the prod server, so using the current academic year might still make sense in practice. Could offer a

> No results. [See the all-time leaderboard.](#)

if people visit the site before the first ACM event of the year

**Pull request merged**